cdap-ui/app/cdap/components/Tags/index.js

```diff
@@ -15,33 +15,123 @@
+import SpotlightModal from 'components/SpotlightSearch/SpotlightModal';
 require('./Tags.scss');
 import T from 'i18n-react';
 
 export default class Tags extends Component {
   constructor(props) {
     super(props);
     this.state = {
-      tags: {
-        system: objectQuery(this.props, 'entity', 'metadata', 'metadata', 'SYSTEM', 'tags') || [],
-        user: objectQuery(this.props, 'entity', 'metadata', 'metadata', 'USER', 'tags') || []
-      }
+      systemTags: [],
+      userTags: [],
+      searchModalOpen: false,
+      showInputField: false,
+      currentInputTag: ''
     };
+
+    this.params = {
+      namespace: NamespaceStore.getState().selectedNamespace,
+      entityType: convertEntityTypeToApi(this.props.entity.type),
+      entityId: this.props.entity.id,
+    };
+
+    this.toggleSearchModal = this.toggleSearchModal.bind(this);
+    this.toggleInputField = this.toggleInputField.bind(this);
+    this.onInputTagChange = this.onInputTagChange.bind(this);
+    this.addTag = this.addTag.bind(this);
+  }
+  componentWillMount() {
+    Mousetrap.bind('enter', this.addTag);
+
+    let systemParams = Object.assign({}, this.params, { scope: 'SYSTEM' });
+    let userParams = Object.assign({}, this.params, { scope: 'USER' });
+
+    MyMetadataApi.getTags(systemParams)
+      .combineLatest(MyMetadataApi.getTags(userParams))
+      .subscribe((res) => {
+        this.setState({
+          systemTags: res[0],
+          userTags: res[1]
+        });
+      }, (err) => {
+        console.log('Error', err);
+      });
   }
-  componentWillReceiveProps(nextProps) {
-    let {entity} = nextProps;
-    if (entity.id !== this.props.entity.id) {
+
+  componentWillUnmount() {
+    Mousetrap.unbind('enter');
+  }
+
+  toggleSearchModal() {
+    this.setState({
+      searchModalOpen: !this.state.searchModalOpen
+    });
+  }
+  toggleInputField() {
+    if (this.state.showInputField) {
       this.setState({
-        tags: {
-          system: objectQuery(nextProps, 'entity', 'metadata', 'metadata', 'SYSTEM', 'tags') || [],
-          user: objectQuery(this.props, 'entity', 'metadata', 'metadata', 'USER', 'tags') || []
-        }
+        currentInputTag: ''
       });
     }
+    this.setState({
+      showInputField: !this.state.showInputField
+    });
+  }
+  onInputTagChange(e) {
+    this.setState({
+      currentInputTag: e.target.value
+    });
+  }
+  fetchUserTags() {
+    let params = Object.assign({}, this.params, { scope: 'USER' });
+
+    MyMetadataApi.getTags(params)
+      .subscribe((res) => {
+        this.setState({
+          userTags: res,
+        });
+      }, (err) => {
+        console.log('Error', err);
+      });
+  }
+  addTag() {
+    if (this.state.currentInputTag !== '') {
+      MyMetadataApi.addTags(this.params, [this.state.currentInputTag])
+      .subscribe(() => {
+        this.fetchUserTags();
+        this.toggleInputField();
+      }, (err) => {
+        console.log(err);
```

Comment:
Same here too. If users are not able to add a tag then we should show appropriate error message for failure.